plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,347 @@
+        SearchScrollRequest request = new SearchScrollRequest(mainRequest);
+        request.scrollId(scroll.get()).scroll(firstSearchRequest.scroll());
+        client.searchScroll(request, new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse response) {
+                onScrollResponse(response);
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                finishHim(e);
+            }
+        });
+    }
+
+    private void recordFailure(Failure failure) {
+        if (failure.getStatus() == CONFLICT) {
+            versionConflicts.incrementAndGet();
+            if (false == mainRequest.abortOnVersionConflict()) {
+                return;
+            }
+        }
+        failures.add(failure);
+    }
+
+    void startNormalTermination() {
+        if (false == mainRequest.refresh()) {
+            finishHim(null);
+            return;
+        }
+        RefreshRequest refresh = new RefreshRequest(mainRequest);
+        refresh.indices(destinationIndices.toArray(new String[destinationIndices.size()]));
+        client.admin().indices().refresh(refresh, new ActionListener<RefreshResponse>() {
+            @Override
+            public void onResponse(RefreshResponse response) {
+                finishHim(null);
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                finishHim(e);
+            }
+        });
+    }
+
+    /**
+     * Finish the request.
+     *
+     * @param failure
+     *            the failure that caused the request to fail prematurely if not
+     *            null. If not null this doesn't mean the request was entirely
+     *            successful - it may have accumulated failures in the failures
+     *            list.
+     */
+    void finishHim(Throwable failure) {
+        String scrollId = scroll.get();
+        if (Strings.hasLength(scrollId)) {
+            /*
+             * Fire off the clear scroll but don't wait for it it return before
+             * we send the use their response.
+             */
+            ClearScrollRequest clearScrollRequest = new ClearScrollRequest(mainRequest);
+            clearScrollRequest.addScrollId(scrollId);
+            client.clearScroll(clearScrollRequest, new ActionListener<ClearScrollResponse>() {
+                @Override
+                public void onResponse(ClearScrollResponse response) {
+                    logger.debug("Freed [{}] contexts", response.getNumFreed());
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    logger.warn("Failed to clear scroll [" + scrollId + ']', e);
+                }
+            });
+        }
+        if (failure == null) {
+            listener.onResponse(buildResponse(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime.get())));
+            return;
+        }
```

Comment:
I can do that. I'm usually a fan of early return but here it doesn't save anything.